pallets/creator-staking/src/types/impls.rs

```diff
@@ -0,0 +1,273 @@
+    /// * `unstake(10, 800)` will result in `[<5, 1000>, <7, 1200>, <9, 800>, <10, 0>]`
+    ///
+    /// # Example 2
+    ///
+    /// `stakes: [<5, 1000>]`
+    /// * `unstake(5, 1000)` will result in `[]`
+    ///
+    /// Note that if no unclaimed eras remain, vector will be cleared.
+    pub(crate) fn unstake(&mut self, current_era: EraIndex, value: Balance) -> Result<(), &str> {
+        if let Some(era_stake) = self.stakes.last_mut() {
+            if era_stake.era > current_era {
+                return Err("Unexpected era")
+            }
+
+            let new_stake_value = era_stake.staked.saturating_sub(value);
+            if current_era == era_stake.era {
+                *era_stake = EraStake::new(new_stake_value, current_era)
+            } else {
+                self.stakes.push(EraStake::new(new_stake_value, current_era))
+            }
+
+            // Removes unstaked values if they're no longer valid for comprehension
+            if !self.stakes.is_empty() && self.stakes[0].staked.is_zero() {
+                self.stakes.remove(0);
+            }
+        }
+
+        Ok(())
+    }
+
+    /// `Claims` the oldest era available for claiming.
+    /// In case valid era exists, returns `(claim era, staked amount)` tuple.
+    /// If no valid era exists, returns `(0, 0)` tuple.
+    ///
+    /// # Example
+    ///
+    /// The following example will demonstrate how the internal vec changes when `claim` is called
+    /// consecutively.
+    ///
+    /// `stakes: [<5, 1000>, <7, 1300>, <8, 0>, <15, 3000>]`
+    ///
+    /// 1. `claim()` will return `(5, 1000)`
+    ///     Internal vector is modified to `[<6, 1000>, <7, 1300>, <8, 0>, <15, 3000>]`
+    ///
+    /// 2. `claim()` will return `(6, 1000)`.
+    ///    Internal vector is modified to `[<7, 1300>, <8, 0>, <15, 3000>]`
+    ///
+    /// 3. `claim()` will return `(7, 1300)`.
+    ///    Internal vector is modified to `[<15, 3000>]`
+    ///    Note that `0` staked period is discarded since nothing can be claimed there.
+    ///
+    /// 4. `claim()` will return `(15, 3000)`.
+    ///    Internal vector is modified to `[16, 3000]`
+    ///
+    /// Repeated calls would continue to modify vector following the same rule as in *4.*
+    pub(crate) fn claim(&mut self) -> (EraIndex, Balance) {
+        if let Some(era_stake) = self.stakes.first() {
+            let era_stake = *era_stake;
+
+            if self.stakes.len() == 1 || self.stakes[1].era > era_stake.era + 1 {
+                self.stakes[0] =
+                    EraStake { staked: era_stake.staked, era: era_stake.era.saturating_add(1) }
+            } else {
+                // in case: self.stakes[1].era == era_stake.era + 1
+                self.stakes.remove(0);
+            }
+
+            // Removes unstaked values if they're no longer valid for comprehension
+            if !self.stakes.is_empty() && self.stakes[0].staked.is_zero() {
```

Comment:
Looks like `!self.stakes.is_empty()` will be always `true`. Let's delete it.
```suggestion
            if self.stakes[0].staked.is_zero() {
```